openmpi: fixes for slurm and #29449

sethrj opened the pull request:
`~pmix` is invalid for openmpi@4: it requires either an external or internal copy, even if pmi is used (contrary to https://github.com/spack/spack/pull/29449#discussion_r824825876 )
```
1 error found in build log:
     171    checking for libpmi2 in /usr/lib... checking for libpmi2 in /usr/lib64... found
     172    checking for PMI2_Init in -lpmi2... yes
     173    checking can PMI support be built... yes
     174    configure: WARNING: Open MPI requires PMIx support. It can be built
     175    configure: WARNING: with either its own internal copy of PMIx, or with
     176    configure: WARNING: an external copy that you supply.
  >> 177    configure: error: Cannot continue
```

At least as of slurm@20, PMI is still required despite https://github.com/spack/spack/pull/29449#discussion_r834670215:
```
The application appears to have been direct launched using "srun",
but OMPI was not built with SLURM's PMI support and therefore cannot
execute. There are several options for building PMI support under
SLURM, depending upon the SLURM version you are using:

  version 16.05 or later: you can use SLURM's PMIx support. This
  requires that you configure and build SLURM --with-pmix.

  Versions earlier than 16.05: you must use either SLURM's PMI-1 or
  PMI-2 support. SLURM builds PMI-1 by default, or you can manually
  install PMI-2. You must then build Open MPI using --with-pmi pointing
  to the SLURM PMI library location.

Please configure as appropriate and try again.
```

Maybe there are some other toggles and dependencies and variants that also work, but requiring PMI for SLURM seems to be the quickest path to getting @haampie 's build working again.

sethrj committed:
openmpi: fixes for slurm and #29449

haampie commented:
isn't it a bit confusing that `openmpi +pmix` now restricts it to an old version, because it's only optional for old versions?

sethrj commented:
That is indeed an unwanted side effect. Think it's better to have a variant that's invariant for new versions? 😬

haampie commented:
I think @sethrj tried to avoid a tri-state without/internal/external for pmix. But maybe that's the better solution @sethrj?

sethrj committed:
openmpi: fix pmix dependencies

sethrj commented:
@hppritcha openmpi@5 seems to be building (pmix built, openmpi configured and is halfway built) with the last patch I pushed.

hppritcha commented:
How are you picking up a newer version of pmix than 4.1.2?

I'm seeing:
```
     2294    checking for pmix.h... yes
     2295    checking for PMIx_Init... yes
     2296    checking version 5x... not found
     2297    configure: WARNING: PRTE does not support PMIx versions
     2298    configure: WARNING: less than v5.0 as it requires access
     2299    configure: WARNING: to the internal PMIx library headers.
  >> 2300    configure: error: Please select a newer version and configure again
     2301    configure: ===== done with 3rd-party/prrte configure =====
  >> 2302    configure: error: PRRTE configuration failed.  Cannot continue.

```

sethrj commented:
With the update, `depends_on('pmix@5:', when='@5:')`, which causes the solver to use `pmix@master` when installing `openmpi@main`:
```
checking version 5x... found
```

We should also switch to an "external" build of prrte rather than an internally built one. This is new with openmpi@5 right?

hppritcha commented:
oh my goodness.  if you want to include external prrte someone needs to do major updates to that package.
I'm finding that pmix@main has a dependency on munge.

hppritcha commented:
yes is the answer to your question about openmpi 5.

hppritcha commented:
I know OSSS OpenSHMEM can use the prrte launch system directly.  I'm sure there may be a few others.  Overlap with spack community may be small however.  @rhc54

rhc54 commented:
At this time, I'm only aware of OMPI v5+ being fully _dependent_ on PRRTE. There are multiple other packages that can and do use it, but they are not dependent on PRRTE. Instead, they are dependent on PMIx - at least optionally, if not fully.

I don't know why you are getting an error message about PRRTE being dependent on PMIx v5 - that isn't correct. It does require at least PMIx v4.0 for PRRTE v2.0, and PRRTE v2.1 may require something like PMIx v4.1 (still TBD as PRRTE v2.1 won't be released for another few months).

Currently, PMIx v4.1.2 is the last release we have made. I'm expecting v4.2 (and a v4.1.3) to come out in another month or so.

hppritcha commented:
the configury error message is being emitted from this m4 code:

[https://github.com/openpmix/prrte/blob/master/config/prte_setup_pmix.m4]( https://github.com/openpmix/prrte/blob/master/config/prte_setup_pmix.m4)

rhc54 commented:
Well yes - but that is on PRRTE's _master_ branch, not a _release_ branch. Why would Spack be publishing our master development branch???

If you look at the latest release (PRRTE v2.0), you'll see:
```m4
                      AC_MSG_WARN([PRTE does not support PMIx versions])
                      AC_MSG_WARN([less than v4.1 as only PMIx-based tools can])
                      AC_MSG_WARN([connect to the server.])
                      AC_MSG_ERROR([Please select a newer version and configure again])])
```
which shows the dependency on PMIx v4.1 and above.

FWIW: my understanding is that OMPI v5 is going to require a minimum of PRRTE v2.1 - which won't be released until probably late summer.

hppritcha commented:
we do use master/main branches for nightly testing, etc. at least I do to give a heads up on problems coming down the road.

rhc54 commented:
> we do use master/main branches for nightly testing, etc. at least I do to give a heads up on problems coming down the road.

I see - okay, I can change the configure requirement. Otherwise, you'd be forced to use PMIx master branch as well, and that could get complicated if you are wanting to retain independence between the two projects. Let me see what the right minimum version would be for PRRTE master.

rhc54 commented:
I've dug into this PMIx v5 constraint a bit more and found the reason behind it. We chose to reduce the code duplication between PRRTE and PMIx, especially in the various utilities they use. This was done by making PRRTE reuse the PMIx equivalents.

However, that in turn requires that PMIx export a number of header files so that PRRTE can build against those utilities. Only the PMIx master branch has been updated to do so - hence the configure constraint to PMIx v5 and above. Note that there is no plan (at least today) to backport the PRRTE/PMIx code duplication changes to any of the existing PRRTE or PMIx release branches.

Moving the utility changes to the PMIx release branches would be a major change to those code bases. I can raise the point at tomorrow's developer meeting (which @hppritcha usually attends) to see how people feel about it.

haampie commented:
What's the status here?

rhc54 commented:
I'm afraid @hppritcha didn't make it to the last PMIx meeting where I had hoped to discuss moving the utility changes back to the release branches. I did, however, send a note out to the mailing list about it - and got no response, which we usually take to mean "nobody cares". Just a question of finding a time to do it.

hppritcha commented:
@rhc54 thanks for looking into the reason for this constraint.  I don't think moving the utility code changes into PMIx release branches is worth the effort.

hppritcha commented:
I would like the openmpi package to use the internal 3rd-party packages by default when building main.

sethrj commented:
I think that's fair for an "experimental" build with rapidly evolving dependency requirements.

sethrj committed:
openmpi: @2 requires pmix@1

```
WARNING: This version of Open MPI does not support PMIx version 2.x and later
```

sethrj committed:
openmpi: always requires pmix

```
     142    configure: WARNING: Open MPI requires PMIx support. It can be built
     143    configure: WARNING: with either its own internal copy of PMIx, or with
     144    configure: WARNING: an external copy that you supply.
```

sethrj committed:
Fix pmix constraint for main branch

sethrj commented:
@hppritcha I've removed the external pmix dependency for `openmpi@main` so it will use the internal one. I also got errors building openmpi@2 with `--disable-pmix` so it looks like pmix is always required (if not provided externally it will be built internally).

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -321,14 +320,17 @@ class Openmpi(AutotoolsPackage, CudaPackage):
+    depends_on('pmix@5:', when='@5.0:5')
+
+    # Libevent is required for PMIx
+    depends_on('libevent', when='@2:')
```
maybe update the comments that "internal" pmix is used when @main?
```suggestion
    depends_on('libevent', when='@main')
```

sethrj replied:
I merged this with a rogue `libevent@2:` so the logic should be correct now

sethrj committed:
Remove transient libevent dependency: only require for vendored pmix

becker33 commented:
This looks like it's all in order to me, I can merge it in the morning to get it into v0.18.0 if no one objects by then

sethrj commented:
Needs reapproval from @haampie

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -747,11 +735,10 @@ def configure_args(self):
             config_args.append(lustre_opt)
         # External libevent/pmix
-        if spec.satisfies('@5:') or spec.satisfies('+pmix'):
+        if '^pmix' in spec:
+            # Externally supplied PMIx
             config_args.append('--with-pmix={0}'.format(spec['pmix'].prefix))
             config_args.append('--with-libevent={0}'.format(spec['libevent'].prefix))
```
```suggestion
        if '^pmix' in spec:
            config_args.append('--with-pmix={0}'.format(spec['pmix'].prefix))
        if '^libevent' in spec:
            config_args.append('--with-libevent={0}'.format(spec['libevent'].prefix))
```

haampie replied:
This will still set `--with-libevent` when pmix is not vendored. I hope OpenMPI ignores the flag since it's not its direct dep?

haampie commented on var/spack/repos/builtin/packages/openmpi/package.py:
```diff
@@ -321,14 +318,17 @@ class Openmpi(AutotoolsPackage, CudaPackage):
+    depends_on('pmix@3.2:', when='@4.0:4')
+    depends_on('pmix@5:', when='@5.0:5')
+
+    # Libevent is required for PMIx
```
```suggestion
    # Libevent is required when vendored PMIx is used
```

haampie reviewed (changes requested):
libevent needs to be fixed

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into openmpi-pmi

sethrj commented:
Good catch @haampie . I've simplified the logic for constructing the arguments and tested it on 4.1.6. The `libevent` argument still gets added as a transient dependency, but it looks from the output as though libevent is still explicitly required by openmpi even when pmix is an external dependency:
```
--- MCA component event:external (m4 configuration macro, priority 90)
checking for MCA component event:external compile mode... static
checking --with-libevent-libdir value... simple ok (unspecified value)
checking for external libevent in... /opt/spack/opt/spack/apple-clang-13.0.0/libevent/73xjwog
checking --with-libevent value... sanity check ok (/opt/spack/opt/spack/apple-clang-13.0.0/libevent/73xjwog)
checking for /opt/spack/opt/spack/apple-clang-13.0.0/libevent/73xjwog/lib64... not found
checking for /opt/spack/opt/spack/apple-clang-13.0.0/libevent/73xjwog/lib... found
checking event2/event.h usability... yes
checking event2/event.h presence... yes
checking for event2/event.h... yes
looking for library in lib
checking for library containing event_config_new... -levent_core
checking for evthread_set_lock_callbacks in -levent_core... yes
checking for evthread_use_pthreads in -levent_pthreads... yes
checking if MCA component event:external can compile... yes
```
so what we have now will always work but it may be more precise to restore `depends_on('libevent', when='@2:')`.

sethrj committed:
Fix dependencies

sethrj commented:
@haampie @becker33 I'm so sorry, I forgot to commit and push this morning after I fixed the dependencies :(
